src/assets/stylesheets/main/_api.scss

```diff
@@ -65,18 +65,36 @@ $objinfo-icon-size: 16px;
         color: var(--md-accent-fg-color);
       }
 
+      // Ensure each parameter starts on a new line and is indented.
       &.sig-wrap {
-        .sig-param {
-          // Ensure each parameter starts on a new line and is indented.
+        .sig-param,
+        .n {
+          + .kt,
```

Comment:
Are you objecting to `.n` or `.kt` or both? I was just going off of what pygments would be looking for.

Reply:
I just mean that it is kind of incidental whether the parameter happens to start with elements that have those pygments styles --- depending on the specific way the parameter is declared, or depending on which language it is, that might or might not work.  It is also not easy for someone reading it to understand what is going on.

Can you remind me why just using `.sig-param` doesn't work here?

Reply:
`sig-param` work for the python domain, but it doesn't work for others (like the C family). In fact, parameters are not always encapsulated in a `sig-param` classed div in the python domain (may have something to do with member's scoping).

Reply:
For Python, if `_parse_arglist` is successful then I believe every parameter will be properly enclosed in a `desc_parameter` docutils node.

However, if the signature cannot be parsed as a real Python signature (e.g. it uses the "[" syntax to indicate optional parts) then Sphinx falls back to `_pseudo_parse_arglist` which does not always enclose arguments in `desc_parameter` nodes.

It looks like the C domain also should enclose each parameter in a `desc_parameter` node, so I'm a bit confused why that isn't working.

https://github.com/sphinx-doc/sphinx/blob/b4276edd848d112b4e981011c334d27cbcb20018/sphinx/domains/c.py#L720

Can you show an example (and e.g. the output of `node.pformat()`) of where that breaks down?  I think it might be better to just fix the domains to produce the proper node structure rather than try to accommodate it in CSS.

Reply:
```rst
.. cpp:function:: const MyType Foo(const MyType bar, uint8_t* arr, unsigned int len = DEFAULT_LENGTH, bool baz= false)
```

<details><summary>output from <code>node.pformat()</code></summary>

```xml
<desc_signature classes="sig sig-object cpp highlight" ids="_CPPv43FooK6MyTypeP7uint8_tjb _CPPv33FooK6MyTypeP7uint8_tjb _CPPv23FooK6MyTypeP7uint8_tjb Foo__MyTypeC.uint8_tP.unsigned-i.b" is_multiline="True">
    <desc_signature_line add_permalink="True" xml:space="preserve">
        <inline classes="k">
            const
        <inline classes="w">

        <reference internal="True" refid="_CPPv46MyType" reftitle="MyType">
            <inline classes="n">
                MyType
        <inline classes="w">

        <desc_name classes="sig-name descname" xml:space="preserve">
            <inline classes="n">
                Foo
        <desc_parameterlist xml:space="preserve">
            <desc_parameter noemph="True" xml:space="preserve">
                <inline classes="k">
                    const
                <inline classes="w">

                <reference internal="True" refid="_CPPv46MyType" reftitle="MyType">
                    <inline classes="n">
                        MyType
                <inline classes="w">

                <inline classes="n sig-param">
                    bar
            <desc_parameter noemph="True" xml:space="preserve">
                <inline classes="n">
                    uint8_t
                <inline classes="w">

                <inline classes="p">
                    *
                <inline classes="n sig-param">
                    arr
            <desc_parameter noemph="True" xml:space="preserve">
                <inline classes="kt">
                    unsigned
                <inline classes="w">

                <inline classes="kt">
                    int
                <inline classes="w">

                <inline classes="n sig-param">
                    len
                <inline classes="w">

                <inline classes="p">
                    =
                <inline classes="w">

                <inline classes="n">
                    DEFAULT_LENGTH
            <desc_parameter noemph="True" xml:space="preserve">
                <inline classes="kt">
                    bool
                <inline classes="w">

                <inline classes="n sig-param">
                    baz
                <inline classes="w">

                <inline classes="p">
                    =
                <inline classes="w">

                <inline classes="k">
                    false
```

</details>

<details><summary>HTML output</summary>

```html
<dt class="sig sig-object cpp highlight sig-wrap" id="_CPPv43FooK6MyTypeP7uint8_tjb">
  <span id="_CPPv33FooK6MyTypeP7uint8_tjb"></span>
  <span id="_CPPv23FooK6MyTypeP7uint8_tjb"></span>
  <span id="Foo__MyTypeC.uint8_tP.unsigned-i.b"></span>
  <span class="k"><span class="pre">const</span></span>
  <span class="w"> </span>
  <a class="reference internal" href="#_CPPv46MyType" title="MyType">
    <span class="n">
      <span class="pre">MyType</span>
    </span>
  </a>
  <span class="w"> </span>
  <span class="sig-name descname">
    <span class="n">
      <span class="pre">Foo</span>
    </span>
  </span>
  <span class="sig-paren">(</span>
  <span class="k">
    <span class="pre">const</span>
  </span>
  <span class="w"> </span>
  <a class="reference internal" href="#_CPPv46MyType" title="MyType">
    <span class="n">
      <span class="pre">MyType</span>
    </span>
  </a>
  <span class="w"> </span>
  <span class="n sig-param">
    <span class="pre">bar</span>
  </span>, 
  <span class="n">
    <span class="pre">uint8_t</span>
  </span>
  <span class="w"> </span>
  <span class="p">
    <span class="pre">*</span>
  </span>
  <span class="n sig-param">
    <span class="pre">arr</span>
  </span>, 
  <span class="kt">
    <span class="pre">unsigned</span>
  </span>
  <span class="w"> </span>
  <span class="kt">
    <span class="pre">int</span>
  </span>
  <span class="w"> </span>
  <span class="n sig-param">
    <span class="pre">len</span>
  </span>
  <span class="w"> </span>
  <span class="p">
    <span class="pre">=</span>
  </span>
  <span class="w"> </span>
  <span class="n">
    <span class="pre">DEFAULT_LENGTH</span>
  </span>, 
  <span class="kt">
    <span class="pre">bool</span>
  </span>
  <span class="w"> </span>
  <span class="n sig-param">
    <span class="pre">baz</span>
  </span>
  <span class="w"> </span>
  <span class="p">
    <span class="pre">=</span>
  </span>
  <span class="w"> </span>
  <span class="k">
    <span class="pre">false</span>
  </span>
  <span class="sig-paren">)</span>
  <a class="headerlink" href="#_CPPv43FooK6MyTypeP7uint8_tjb" title="Permalink to this definition">¶</a>
  <br />
</dt>
<dd>
  <p>Some function type thing</p>
</dd>
```

</details>

---

```rst
.. py:function:: func(long: int, param: str, args: None, flags: bool, lists: Union[list, tuple])
```

<details><summary>output from <code>node.pformat()</code></summary>

```xml
<desc_signature class="" classes="sig sig-object py highlight" fullname="func" ids="test_py_module.test.func" module="test_py_module.test">
    <desc_addname classes="sig-prename descclassname" xml:space="preserve">
        test_py_module.test.
    <desc_name classes="sig-name descname" xml:space="preserve">
        func
    <desc_parameterlist xml:space="preserve">
        <desc_parameter xml:space="preserve">
            <inline classes="n">
                long
            <inline classes="p">
                :
            <inline classes="w">

            <inline classes="n">
                <desc_type xml:space="preserve">
                    <reference internal="False" reftitle="(in Python v3.10)" refuri="https://docs.python.org/3/library/functions.html#int">
                        int
        <desc_parameter xml:space="preserve">
            <inline classes="n">
                param
            <inline classes="p">
                :
            <inline classes="w">

            <inline classes="n">
                <desc_type xml:space="preserve">
                    <reference internal="False" reftitle="(in Python v3.10)" refuri="https://docs.python.org/3/library/stdtypes.html#str">
                        str
        <desc_parameter xml:space="preserve">
            <inline classes="n">
                args
            <inline classes="p">
                :
            <inline classes="w">

            <inline classes="n">
                <desc_type xml:space="preserve">
                    <reference internal="False" reftitle="(in Python v3.10)" refuri="https://docs.python.org/3/library/constants.html#None">
                        None
        <desc_parameter xml:space="preserve">
            <inline classes="n">
                flags
            <inline classes="p">
                :
            <inline classes="w">

            <inline classes="n">
                <desc_type xml:space="preserve">
                    <reference internal="False" reftitle="(in Python v3.10)" refuri="https://docs.python.org/3/library/functions.html#bool">
                        bool
        <desc_parameter xml:space="preserve">
            <inline classes="n">
                lists
            <inline classes="p">
                :
            <inline classes="w">

            <inline classes="n">
                <desc_type xml:space="preserve">
                    <reference internal="False" reftitle="(in Python v3.10)" refuri="https://docs.python.org/3/library/typing.html#typing.Union">
                        Union
                    <inline classes="p">
                        [
                    <reference internal="False" reftitle="(in Python v3.10)" refuri="https://docs.python.org/3/library/stdtypes.html#list">
                        list
                    <inline classes="p">
                        ,
                    <inline classes="w">

                    <reference internal="False" reftitle="(in Python v3.10)" refuri="https://docs.python.org/3/library/stdtypes.html#tuple">
                        tuple
                    <inline classes="p">
                        ]
    <reference internal="True" refuri="_modules/test_py_module/test.html#func">
        <inline classes="viewcode-link">
            [source]
```

</details>

<details><summary>HTML output</summary>

```html
<dt class="sig sig-object py highlight sig-wrap" id="test_py_module.test.func">
  <span class="sig-prename descclassname">
    <span class="pre">test_py_module.test.</span>
  </span>
  <span class="sig-name descname">
    <span class="pre">func</span>
  </span>
  <span class="sig-paren">(</span>
  <em class="sig-param">
    <span class="n">
      <span class="pre">long</span>
    </span>
    <span class="p">
      <span class="pre">:</span>
    </span>
    <span class="w"> </span>
    <span class="n">
      <span class="desctype">
        <a class="reference external" href="https://docs.python.org/3/library/functions.html#int"
          title="(in Python v3.10)">
          <span class="pre">int</span>
        </a>
      </span>
    </span>
  </em>,
  <em class="sig-param">
    <span class="n">
      <span class="pre">param</span>
    </span>
    <span class="p">
      <span class="pre">:</span>
    </span>
    <span class="w"> </span>
    <span class="n">
      <span class="desctype">
        <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str"
          title="(in Python v3.10)">
          <span class="pre">str</span>
        </a>
      </span>
    </span>
  </em>,
  <em class="sig-param">
    <span class="n">
      <span class="pre">args</span>
    </span>
    <span class="p">
      <span class="pre">:</span>
    </span>
    <span class="w"> </span>
    <span class="n">
      <span class="desctype">
        <a class="reference external" href="https://docs.python.org/3/library/constants.html#None"
          title="(in Python v3.10)">
          <span class="pre">None</span>
        </a>
      </span>
    </span>
  </em>,
  <em class="sig-param">
    <span class="n">
      <span class="pre">flags</span>
    </span>
    <span class="p">
      <span class="pre">:</span>
    </span>
    <span class="w"> </span>
    <span class="n">
      <span class="desctype">
        <a class="reference external" href="https://docs.python.org/3/library/functions.html#bool"
          title="(in Python v3.10)">
          <span class="pre">bool</span>
        </a>
      </span>
    </span>
  </em>,
  <em class="sig-param">
    <span class="n">
      <span class="pre">lists</span>
    </span>
    <span class="p">
      <span class="pre">:</span>
    </span>
    <span class="w"> </span>
    <span class="n">
      <span class="desctype">
        <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Union"
          title="(in Python v3.10)">
          <span class="pre">Union</span>
        </a>
        <span class="p">
          <span class="pre">[</span>
        </span>
        <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list"
          title="(in Python v3.10)">
          <span class="pre">list</span>
        </a>
        <span class="p">
          <span class="pre">,</span>
        </span>
        <span class="w"> </span>
        <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple"
          title="(in Python v3.10)">
          <span class="pre">tuple</span>
        </a>
        <span class="p">
          <span class="pre">]</span>
        </span>
      </span>
    </span>
  </em>
  <span class="sig-paren">)</span>
  <a class="reference internal" href="_modules/test_py_module/test.html#func">
    <span class="viewcode-link">
      <span class="pre">[source]</span>
    </span>
  </a>
  <a class="headerlink" href="#test_py_module.test.func" title="Permalink to this definition">¶</a>
</dt>
<dd>
  <p>A function with many parameters.</p>
</dd>
```

</details>

---

```rst
.. py:function:: format_exception(etype, value, tb[, limit=None])
```

<details><summary>output from <code>node.pformat()</code></summary>

```xml
<desc_signature class="" classes="sig sig-object py highlight" fullname="format_exception" ids="format_exception" module="True">
    <desc_name classes="sig-name descname" xml:space="preserve">
        format_exception
    <desc_parameterlist xml:space="preserve">
        <desc_parameter xml:space="preserve">
            <inline classes="n">
                etype
        <desc_parameter xml:space="preserve">
            <inline classes="n">
                value
        <desc_parameter xml:space="preserve">
            <inline classes="n">
                tb
        <desc_optional xml:space="preserve">
            <desc_parameter xml:space="preserve">
                <inline classes="n">
                    limit=None
```

</details>

<details><summary>HTML output</summary>

```html
<dt class="sig sig-object py highlight" id="format_exception">
  <span class="sig-name descname">
    <span class="pre">format_exception</span>
  </span>
  <span class="sig-paren">(</span>
  <em class="sig-param">
    <span class="n">
      <span class="pre">etype</span>
    </span>
  </em>,
  <em class="sig-param">
    <span class="n">
      <span class="pre">value</span>
    </span>
  </em>,
  <em class="sig-param">
    <span class="n">
      <span class="pre">tb</span>
    </span>
  </em>
  <span class="optional">[</span>,
  <em class="sig-param">
    <span class="n">
      <span class="pre">limit=None</span>
    </span>
  </em>
  <span class="optional">]</span>
  <span class="sig-paren">)</span>
  <a class="headerlink" href="#format_exception" title="Permalink to this definition">¶</a>
</dt>
<dd>
  <p>Format the exception with a traceback.</p>
  <dl class="field-list simple">
    <dt class="field-odd">Parameters</dt>
    <dd class="field-odd">
      <dl class="api-field simple">
        <dt><span class="sig-name descname">etype</span></dt>
        <dd>
          <p>exception type</p>
        </dd>
        <dt><span class="sig-name descname">value</span></dt>
        <dd>
          <p>exception value</p>
        </dd>
        <dt><span class="sig-name descname">tb</span></dt>
        <dd>
          <p>traceback object</p>
        </dd>
        <dt>
          <span class="sig-name descname">limit</span>
          <span class="p"> : </span>
          <span class="desctype">
            <span class="desctype">integer or None</span>
          </span>
        </dt>
        <dd>
          <p>maximum number of stack frames to show</p>
        </dd>
      </dl>
    </dd>
    <dt class="field-even">Return type</dt>
    <dd class="field-even">
      <p>list of strings</p>
    </dd>
  </dl>
</dd>
```

</details>

Reply:
Thanks for gathering all of that --- that is really helpful for understanding what is going on.

So it looks like for the C and C++ domains we can count on every parameter being contained in a desc_parameter node, and for Python we mostly can as well as long as we don't care about the not-real-Python optional syntax.

If `desc_parameter` resulted in `<span class="sig-param">` then we would be all set.  Unfortunately, that is not always the case:

In fact `desc_parameter` only results in a `sig-param` class when using the html5 writer (which our theme always uses) and additionally when the `noemph` attribute is not set:
https://github.com/sphinx-doc/sphinx/blob/b4276edd848d112b4e981011c334d27cbcb20018/sphinx/writers/html5.py#L178

The `noemph` attribute is set by the C and C++ domains on all of the `desc_parameter` nodes that they create, so we never get `sig-param` classes on the `desc_parameter` nodes:
https://github.com/sphinx-doc/sphinx/blob/b4276edd848d112b4e981011c334d27cbcb20018/sphinx/domains/c.py#L720

In contrast the Python domain never sets `noemph=True` so Python parameters are always wrapped in `<em class="sig-param">`.

We can see though that the  C++ domain does still output `class="sig-param"` on the inner name portion of the parameter.  That is because it manually adds that class:
https://github.com/sphinx-doc/sphinx/blob/b4276edd848d112b4e981011c334d27cbcb20018/sphinx/domains/cpp.py#L650

In summary, we clearly can't rely on the `sig-param` class as provided by the existing HTML output for much of anything.

I think we should just add custom `visit_desc_parameter` and `depart_desc_parameter` methods here:
https://github.com/jbms/sphinx-immaterial/blob/4b7f27ae05542cbcca979be4f0dfe905d78f141b/sphinx_immaterial/apidoc_formatting.py#L167

that will wrap every parameter in:

`<span class="sig-param-decl">`

Then we can change our CSS rules to reference this new `sig-param-decl` class rather than `sig-param`.  We can also have `visit_desc_parameter` set `noemph=True` so that we don't get the redundant `<em class="sig-param">` also wrapping Python parameters.

We would no longer have any style rules for `sig-param` so the extra class added by the C++ domain would have no effect.


However, I see that `desc_parameter` does not always result in a `sig-param` class (added by sphinx/writers/html5.py), as that depends on the `noemph` property not being set.  The C++ domain always sets `noemph`, so we never get a sig-param class.

Reply:
Great analysis! I've locally tested the `visit/depart_desc_parameter()` idea with _api.scss adjusted accordingly, and it worked like a charm. Committed solution for review.

> We can also have visit_desc_parameter set noemph=True so that we don't get the redundant `<em class="sig-param">` also wrapping Python parameters.

I opted not to do this since I wasn't sure how to determine node's domain. Since we aren't using the `sig-param` class in CSS anymore, it seemed redundant to modify the original behavior.

### Side-effects
- All parameters (albeit multiple or single line signature) will be wrapped in a `span.sig-param-decl` element. So users could extend the multi-line approach to single line signatures if they so choose.
- The non-pythonic optional syntax is not wrapped with the `*_desc_parameter()`
   ```html
   <span class="sig-name descname"><!--  --></span>
   <span class="sig-paren">(</span>
   <span class="sig-param-decl"><!-- -->, </span>
   <span class="optional">[</span>
   <span class="sig-param-decl"><!-- -->, </span>
   <span class="optional">]</span>
   <span class="sig-paren">)</span>
   ```
   This ends up looking a little weird for multi-lined signatures, but not at all alien:
   ![image](https://user-images.githubusercontent.com/14963867/163387036-6488b7d3-df96-4933-85e5-8fdaaa901f33.png)

Reply:
yeah the "optional" syntax doesn't look great but I think we can just encourage users not to use it